Nodejs/Setup/InstallerTelemetryAction/InstallerTelemetryAction.cs

```diff
@@ -0,0 +1,313 @@
+﻿//*********************************************************//
+//    Copyright (c) Microsoft. All rights reserved.
+//    
+//    Apache 2.0 License
+//    
+//    You may obtain a copy of the License at
+//    http://www.apache.org/licenses/LICENSE-2.0
+//    
+//    Unless required by applicable law or agreed to in writing, software 
+//    distributed under the License is distributed on an "AS IS" BASIS, 
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or 
+//    implied. See the License for the specific language governing 
+//    permissions and limitations under the License.
+//
+//*********************************************************//
+
+using System;
+using System.Collections.Generic;
+using System.Text;
+using Microsoft.Deployment.WindowsInstaller;
+using System.Diagnostics;
+using System.Net;
+using System.Collections.Specialized;
+using System.Web.Script.Serialization;
+using System.Globalization;
+
+namespace Microsoft.NodejsTools
+{
+    public class InstallerTelemetryActions
+    {
+        private static string AIVersion = "2";
+        [CustomAction]
+        public static ActionResult RecordInstallStartTime(Session session)
+        {
+            session["InstallStartTime"] = DateTime.Now.ToString(CultureInfo.InvariantCulture);
+            return ActionResult.Success;
+        }
+
+        [CustomAction]
+        public static ActionResult LogInstallSuccessResult(Session session)
+        {
+            TimeSpan installTime = DateTime.Now - DateTime.Parse(session["InstallStartTime"]);
+            session.Log("Begin Telemetry Log");
+            bool isInstalled = session.EvaluateCondition("Installed");
+            string currentState = "";
+            string requestState = "";
+
+            FeatureInfoCollection fc = session.Features;
+            foreach (FeatureInfo f in fc)
+            {
+                currentState = f.CurrentState.ToString();
+                requestState = f.RequestState.ToString();
+                // we just want the current and requested state of A feature to understand if its a new user, upgrade, reinstall or remove.
+                break; 
+            }
+
+            session.Log("Starting POST");
+            Dictionary<string, object> data = new Dictionary<string, object>()

```

Comment:
As far as I can tell, these objects are structured as follows:
* iKey (constant)
* name (constant)
* time (current time)
* data
  * baseType (constant)
  * baseData
    * ver (constant)
    * name (constant)
    * properties
      * InstallStatus (string)
      * IsNtvsInstalled (based on session)
      * CurrentState (based on session)
      * RequestState (based on session)
      * NtvsVersion (based on session)
      * VSVersion (based on session)
      * MsiVersion (based on session)
      * TimeTakenInSeconds (based on session)

So you can just extract most of this logic into a single method ```LogInstallStatus(installStatus, session)``` and make things way more readable and maintainable. 

Also seeing these all together now, it's odd that the properties are PascalCased, whereas everything else is camelCased.

Also once you refactor this, defining constants will no longer be necessary because everything will be in one place.